core/src/main/scala/cats/Foldable.scala

```diff
@@ -78,11 +80,33 @@ import simulacrum.typeclass
 
   /**
    * Left associative monadic folding on `F`.
+   * Stack safe, not short-circuiting (always folds the whole structure).
    */
-  def foldM[G[_], A, B](fa: F[A], z: B)(f: (B, A) => G[B])(implicit G: Monad[G]): G[B] =
+  def foldLeftM[G[_], A, B](fa: F[A], z: B)(f: (B, A) => G[B])(implicit G: Monad[G]): G[B] =
     foldLeft(fa, G.pure(z))((gb, a) => G.flatMap(gb)(f(_, a)))
 
   /**
+   * Perform a stack-safe monadic fold from the source context `F`
+   * into the target monad `G`.
+   *
+   * This method can express short-circuiting semantics. Even when
+   * `fa` is an infinite structure, this method can potentially
+   * terminate if the `foldRight` implementation for `F` and the
+   * `tailRecM` implementation for `G` are sufficiently lazy.
+   *
+   * Instances for concrete structures (e.g. `List`) will often
+   * have a more efficient implementation than the default one
+   * in terms of `foldRight`.
+   */
+  def foldLeftMRec[G[_], A, B](fa: F[A], z: B)(f: (B, A) => G[B])(implicit G: MonadRec[G]): G[B] = {
```

Comment:
can we rename this `foldLeftM` or just stick with `foldM`. I'm not sure we should be changing this name this late in the game, although I agree `foldLeftM` is a better name.

Maybe add a deprecated final `foldM` that redirects to `foldLeftM`?